### PR TITLE
feat: add parent_id to allow for task trees

### PR DESCRIPTION
**Pull request by AmanuelAaron**
## Checklist

- [ ] Changes have been manually QA'd
- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
  See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.

## Ticket
DET-10016

### CODE REVIEW

File: master/static/migrations/20231206140233_add-parent-id-task-config.tx.up.sql [hunk 1]
```diff
@@ -0,0 +1,2 @@
+ALTER TABLE tasks
+ADD parent_id text REFERENCES tasks(task_id) ON DELETE CASCADE DEFAULT(NULL);
```
Comment (AmanuelAaron) [inline master/static/migrations/20231206140233_add-parent-id-task-config.tx.up.sql:2]: I assume we want child tasks to be deleted when parent tasks are. Doing this through table definition seem efficient. Thought?

File: master/pkg/model/task.go [hunk 2]
```diff
@@ -79,7 +79,8 @@ type Task struct {
 	LogVersion TaskLogVersion `db:"log_version"`
 
 	// Relations.
-	Job *Job `bun:"rel:belongs-to,join:job_id=job_id"`
+	Job      *Job    `bun:"rel:belongs-to,join:job_id=job_id"`
+	ParentID *string `db:"parent_id`
```
Comment (stoksc) [inline master/pkg/model/task.go:83]: isn't the parent a task, always? if some shouldn't be type it as a `TaskID`?

File: master/static/migrations/20231206140233_add-parent-id-task-config.tx.up.sql (hunk 1 above)
Comment (stoksc) [reply master/static/migrations/20231206140233_add-parent-id-task-config.tx.up.sql:2]: yeah, I think this makes sense.

**Review by stoksc — COMMENTED** [review]
lgtm, modulo making parent_id a TaskID.

File: master/internal/api_generic_tasks.go [hunk 3]
```diff
@@ -205,6 +205,7 @@ func (a *apiServer) CreateGenericTask(
 			JobID:      &jobID,
 			LogVersion: model.CurrentTaskLogVersion,
 			Config:     ptrs.Ptr(string(configBytes)),
+			ParentID:   ptrs.Ptr(model.TaskID(*req.ParentId)),
```
Comment (ioga) [inline master/internal/api_generic_tasks.go:208]: won't `req.ParentId` be null for root tasks, causing their to panic on a nil pointer dereference?

Comment (AmanuelAaron) [reply master/internal/api_generic_tasks.go]: Right! fixed

**Review by ioga — APPROVED** [review]

File: master/internal/api_generic_tasks.go [hunk 4]
```diff
@@ -198,13 +198,18 @@ func (a *apiServer) CreateGenericTask(
 			return fmt.Errorf("handling experiment config %v: %w", genericTaskSpec.GenericTaskConfig, err)
 		}
 
+		var parentTaskID *model.TaskID
+		if req.ParentId != nil {
+			parentTaskID = ptrs.Ptr(model.TaskID(*req.ParentId))
```
Comment (stoksc) [inline master/internal/api_generic_tasks.go:203]: i think there is a better way to do this

Comment (stoksc) [inline master/internal/api_generic_tasks.go:203]:
```suggestion
			parentTaskID = (*model.TaskID)(req.ParentId)
```

**Pull request closed without merging**